k2/csrc/fsa_algo.cc

```diff
@@ -231,7 +228,7 @@ bool Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
   bool is_acyclic = ConnectCore(a, &state_b_to_a);
   if (state_b_to_a.empty()) return true;
 
-  b->arc_indexes.resize(state_b_to_a.size());
+  b->arc_indexes.resize(state_b_to_a.size() + 1);
```

Comment:
You do not need to handle the case for `if`, but you have to deal the case for `resize`.

Do we really need to put an extra item inside `arc_indexes`? It is error-prone.

Reply:
Actually here's one issue of implentation as you get `b.numstates()` here and try to use it in below code.
https://github.com/danpovey/k2/blob/aa79c3b42a4df2675878bc8782d25d8427d5c7f5/k2/csrc/fsa_algo.cc#L242
I suggest we should avoid to get `target.numstates` and use it. Instead, we should just `push_back` (instead assign) a duplicate at the end of any algorithms, so all implementations will keep the same.